### PR TITLE
WIP - Remove publish with ack

### PR DESCRIPTION
**Pull request by mneumann**
Related PR #299. This is an alternative approach that does not need any modifications of rumqttc. Completely untested.

### CODE REVIEW

**Review by makr11st — COMMENTED** [review]
I would like to test it, but to be honest I like this approach. I think we don't loose anything by not customizing MQTT protocol and the lib but we gain a lot by not having to maintaining our own fork.

**Review by didier-wenzek — COMMENTED** [review]
This proposal is the way to go. As noted by Lukazs, the `publish_with_ack` method is used only by `tedge mqtt pub`. Hence, the price of having [our own fork](#299) is quite huge compared to the benefits.

However, the logic using a pending count of messages needs to be clarified/simplified/fixed.
* `tedge mqtt pub foo bar --qos 2` is no more working (QoS 0 and 1 are working).
* It's not clear why we need a `pending` map and a `pending_count`.
* We do no more need to store the pending messages.
* I was expecting some count decr on PubAck and PubRel. Did I miss something?

File: common/mqtt_client/src/lib.rs [hunk 1]
```diff
@@ -184,17 +184,17 @@ impl Client {
...
-                                .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
                         }
                     }
                 }
 
+                Ok(Event::Incoming(Packet::PubAck(_))) => {
+                    pending_count
+                        .as_ref()
+                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                }
```
Comment (didier-wenzek) [inline common/mqtt_client/src/lib.rs:196]: Something similar needs to be done for QoS 2 on reception of a [PubComp](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718058) and not of a PubRel that are related to *incoming* messages.

```suggestion
                Ok(Event::Incoming(Packet::PubAck(_))) => {
                    pending_count
                        .as_ref()
                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
                }

                Ok(Event::Incoming(Packet::PubComp(_))) => {
                    pending_count
                        .as_ref()
                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
                }
```

File: common/mqtt_client/src/lib.rs [hunk 2]
```diff
@@ -239,16 +184,22 @@ impl Client {
...
+
                 Ok(Event::Incoming(Packet::PubRel(pubrel))) => {
                     if let Some(msg) = pending.remove(&pubrel.pkid) {
+                        pending_count
+                            .as_ref()
+                            .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
```
Comment (didier-wenzek) [inline common/mqtt_client/src/lib.rs:202]: This must be removed. Indeed, the `PubRel` message is related to received messages, while the `pending_count` is related to published messages.

```suggestion
```

**Review by didier-wenzek — COMMENTED** [review]
With the suggested changes the code is working fine in the 3 following cases:

```
$ tedge mqtt pub foo bar --qos 0
$ tedge mqtt pub foo bar --qos 1
$ tedge mqtt pub foo bar --qos 2
```

I also suggest to rename:
* `pending` to `pending_received_messages`,
* `pending_count` to `pending_published_count`

File: common/mqtt_client/src/lib.rs [hunk 3]
```diff
@@ -257,6 +258,14 @@ impl MqttClient for Client {
             .await
             .map_err(|err| Error::ClientError(rumqttc::ClientError::Request(err)))?;
 
+        match qos {
+            QoS::AtMostOnce | QoS::ExactlyOnce => {
```
[the remark this answers is not in the document]
Comment (mneumann) [reply common/mqtt_client/src/lib.rs:262]: Argh, of course :/

**Review by makr11st — APPROVED** [review]
Works for me without issues.

**Review by didier-wenzek — APPROVED** [review]
Approved

**Comment by mneumann** [comment]
@didier-wenzek @makr11st rebased against main. Had some fallout due to the renaming of Error -> MqttClientError. Will merge once CI succeeds. Still would be glad if someone of you could still run simple tests (unless we have automated tests for that).

**Pull request merged**